keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1657 @@
+not needed yet.
+
+To support arbitrarily complex parameters, both ResourceClass and
+ResourceClaim contain one field which holds a
+[runtime.RawExtension](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension) (see [user stories](#user-stories) for examples).
+Validation can be handled by drivers through an
+admission controller (if desired) or later at runtime when the
+parameters are passed to the driver. 
+
+The ResourceClaim spec is immutable. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost, which in the past was not ruled
+out. For example, status could have been stored in a separate etcd instance
+with lower reliability. To recover after a loss, status was meant to be recoverable.
+A [recent KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/2527-clarify-status-observations-vs-rbac)
+clarified that status will always be stored reliably and can be used as
+proposed in this KEP.
+
+Kube-scheduler
+and resource driver communicate by modifying that status. The status is also
+how Kubernetes tracks that a driver has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the
+PersistentVolume/PersistentVolumeClaim model for volumes where the additional
+PV object was used to capture status. That model allowed operations like
+pre-provisioning volumes and then having Kubernetes bind those to claims that
+get created later. For resources, the resource driver can and must handle such
+pre-provisioning internally. Kubernetes wouldn't know how to match
+pre-provisioned resources against claims because it has no understanding about the
+parameters.
+
+Allocation of a resource happens either immediately when a ResourceClaim gets
+created (“immediate allocation”) or when when a Pod is getting scheduled which
+needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec. Pods reference resource
+claims in a new PodSpec.Resources list. Each resource in that list
+then can be made available to one or more containers in that Pod.
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
+downside is that Pod resource requirements cannot be considered when choosing
+where to allocate. If a resource was allocated so that it is only available on
+one node and the Pod cannot run there because other resources like RAM or CPU
+are exhausted on that node, then the Pod cannot run elsewhere. The same applies
+to resources that are available on a certain subset of the nodes and those
+nodes are busy.
+Delayed allocation solves this by integrating allocation with Pod scheduling.
+
+Allocation must be complete before a Pod is allowed to be scheduled onto a
```

Comment:
I don't understand this.  If the device is being lazy-allocated at scheduling time, how can "Allocation must be complete before a Pod is allowed to be scheduled" be satisfied?  I feel like I am missing some high-level overview of what is happening here